src/OrleansRuntime/MembershipService/MembershipOracleData.cs

```diff
@@ -271,6 +300,76 @@ private void NotifyLocalSubscribers(SiloAddress siloAddress, SiloStatus newStatu
             }
         }
 
+        private struct UpdateFaultCombo : IComparable<UpdateFaultCombo>
+        {
+            public int UpdateZone;
+            public int FaultZone;
+        
+            public int CompareTo(UpdateFaultCombo other)
+            {
+ 	             var i = UpdateZone.CompareTo(other.UpdateZone);
+                 return (i != 0) ? i : FaultZone.CompareTo(other.FaultZone);
+           }
+        }
+    
+
+        // deterministic function for designating the silos that should act as gateways
+        private List<SiloAddress> DetermineMultiClusterGateways()
+        {
+            Debug.Assert(!string.IsNullOrEmpty(GlobalServiceId)); // call only if this is a multi cluster
+
+            var candidates = localTableCopyOnlyActive.Keys.ToList();
+
+            // sort deterministically so everyone ends up with the same result
+            candidates.Sort((a, b) =>  
```

Comment:
I didn't think sorting was expensive at all - do you have experience to the contrary?
(yes, it is not linear, but is only O(n lg n)). If necessary we can optimize this further. This piece of code is not that important either. It deterministically selects a subset of nodes to serve as multi-cluster gateways. If you don't want to do that, just choose a number of gateways that is equal to or larger than the number of silos, then you bypass all of this.